core/src/main/java/com/linecorp/armeria/server/ServiceRequestContext.java

```diff
@@ -281,6 +281,18 @@ default MediaType negotiatedProduceType() {
      */
     void setRequestTimeout(Duration requestTimeout);
```

Comment:
We need to improve the javadoc so it makes clearer what these do, using wording about the start time of a request. It would be good to have use case examples too. I might also order the after methods first before absolute ones I'm pretty sure they are more normally used

Reply:
That is a nice idea. I will update Javadoc and examples.

Reply:
I wonder if we need some enum for clarity:

```java
ctx.setRequestTimeout(TimeoutBasis.NOW, Duration.ofSeconds(10));
ctx.setRequestTimeout(TimeoutBasis.DEADLINE, Duration.ofSeconds(10));
ctx.setRequestTimeout(TimeoutBasis.START, Duration.ofSeconds(10));
```

Reply:
That sounds good to me. But I guess we need to change our internal api to support `TimeoutBasis.NOW` to update `SeviceRequestContext.requestTimeoutMilles` correctly.

Reply:
> ctx.setRequestTimeout(TimeoutBasis.DEADLINE, Duration.ofSeconds(10));

Based on our discussion(https://github.com/line/armeria/pull/2343/files#r361444153), `TimeoutBasis.DEADLINE` seems not to go with `Duration`. It needs absolute time like nanoseconds.

Reply:
`DEADLINE` here means `new deadline = deadline + specified value`.

Reply:
> Good morning. 🌞 This API makes senses after sleeping. This means that it extends 10 seconds response timeout from the existing deadline. 😅

This was in pending review... 😂

Reply:
I agree that less options is better. I'm worried that even with two users won't understand well and pick `START` when they wanted `NOW`.

Personally, I haven't come up with a good use case for `START` either and like most the idea of keeping it to just relative timeout. I guess it could be something like retrieving the timeout from an external service? But even then I think most users would expect the timeout to apply to the following business logic and not an invisible request start time.

If we can come up with good javadoc including use case examples for each option I think it's fine to have, or otherwise we could probably remove the option. `DEADLINE` in particular indeed seems hard to come up with ideas for.

There's also an option of having only `NOW` and another method for setting an absolute time as the deadline. The user would have flexibility to use whatever time basis they want and I think it's easier to explain than using `START` or `DEADLINE` in our API.

This is similar to gRPC which internally only has deadline and provides helpers for creating a new deadline from current time or arbitrary time.

https://grpc.github.io/grpc-java/javadoc/io/grpc/Deadline.html

Reply:
> If I implement a streaming service, I guess `TimeoutBasis.NOW` and `TimeoutBasis.START` is enough for me in most cases. Because a user should keep track of the previous status to use `TimeoutBasis.DEADLINE`.

There's a good use case for `DEADLINE` in Central Dogma, where a watch service extends timeout by the amount specified by a user. A user does not need to keep track of the previous status.

----

I think `START` and `DEADLINE` can be merged into one option because a user can calculate one from another: `newTimeout = oldTimeout + adjustment`

So, how about this:

```java
ctx.adjustRequestTimeout(Duration adjustment);
ctx.restartRequestTimeout(Duration newTimeout);
```

Reply:
> I think START and DEADLINE can be merged into one option because a user can calculate one from another: newTimeout = oldTimeout + adjustment

I love this idea. ❤️ 

> ctx.adjustRequestTimeout(Duration adjustment);
ctx.restartRequestTimeout(Duration newTimeout);

Let me use this method naming. :-)

Reply:
I slightly prefer `reset` than `restart`. 😄